When using a C++ struct as a parameter, map its type to a Carbon class type

bricknerb opened the pull request:
This doesn't support actually passing the value of the struct, which is planned to be implemented using thunks.

`ClangDeclId` value is now `ClangDecl` which includes the mapped Carbon instruction in addition to the Clang declaration. This allows finding the Carbon instruction for a given Clang declaration, which is necessary for mapping a Clang struct parameter type to the Carbon class without doing name lookup. We don't take the instruction as part of the hash key, as discussed in [Discord](https://discord.com/channels/655572317891461132/768530752592805919/1380575881050718469).

To map the type, we also need to map namespaces. To avoid recursion for inner namespaces, we use a vector.

Note that the first commit just changes the order of functions in the file to make review easier.

C++ Interop Demo (that shows missing behavior):

```c++
// hello_world.h

struct S {
  S(const S&) { x = 1; }
  int x;
};

void hello_world(S s);
```

```c++
// hello_world.cpp

#include "hello_world.h"

#include <cstdio>

void hello_world2(S s) { printf("hello_world2: %d\n", s.x); }

void hello_world(S s) {
  printf("hello_world: %d\n", s.x);
  hello_world2(s);
}
```

```carbon
// main.carbon

library "Main";

import Cpp library "hello_world.h";

fn Run() -> i32 {
  var s : Cpp.S;
  Cpp.hello_world(s);
  return 0;
}
```

```shell
$ clang -c hello_world.cpp
$ bazel-bin/toolchain/carbon compile main.carbon
$ bazel-bin/toolchain/carbon link hello_world.o main.o --output=demo
$ ./demo
hello_world: -1108224096
hello_world2: 1
```

Part of #5533.

geoffromer commented on toolchain/check/import_cpp.cpp:
```diff
@@ -605,6 +661,9 @@ static auto ImportCXXRecordDecl(Context& context, SemIR::LocId loc_id,
                          // TODO: Set block.
                          /*body=*/{});
 
+  CompleteTypeOrCheckFail(context,
```
If I understand correctly, you added this to prevent a crash that looks like this:

```
CHECK failure at toolchain/check/pattern_match.cpp:379: static_cast<size_t>(param_pattern.index.index) == results_.size(): Parameters out of order; expecting 0 but got -1
```

If so, I think that crash indicates violation of an invariant that `CalleePatternMatch` is supposed to establish, but the import logic isn't calling `CalleePatternMatch`. Instead `MakeParamPatternsBlockId` is trying to emulate it, and there are some bugs in the emulation (see also [here](https://github.com/carbon-language/carbon-lang/issues/5449#issuecomment-2892469303)). I'd strongly encourage you to use `CalleePatternMatch` whenever you're creating an `EntityWithParams`. If there are major obstacles to doing that, let me know; I'm happy to help figure out a solution.

bricknerb replied:
Thanks!
It seems like `CalleePatternMatch()` would be called as part of resolving #5449, so I'll hold off this change until we manage to do that. I know @ivanaivanovska has been working on this as her top priority.

bricknerb replied:
Done.

ivanaivanovska commented on toolchain/check/testdata/interop/cpp/no_prelude/function.carbon:
```diff
@@ -152,26 +152,110 @@ fn F() {
 auto foo(S) -> void;
 
-// --- fail_todo_import_struct_param_type.carbon
+// --- todo_import_struct_definition_param_type.carbon
```
remove `todo_`?

bricknerb replied:
Done.

ivanaivanovska commented on toolchain/check/import_cpp.cpp:
```diff
@@ -347,11 +345,60 @@ static auto MapType(Context& context, clang::QualType type) -> TypeExpr {
 
+// Maps a C++ record type to a Carbon type.
+// TODO: Support more record types.
+static auto MapRecordType(Context& context, SemIR::LocId loc_id,
```
Can we add a test for the scope? (e.g. the name not found in the scope)

bricknerb replied:
We no longer do name lookup.

bricknerb committed:
Change function order without any changes to make logic changes easier to review

bricknerb committed:
Map a C++ struct function parameter to a Carbon class
`ClangDeclId` value is `ClangDecl` which includes the mapped Carbon instruction in addition to the Clang declaration. This allows finding the Carbon instruction for a given Clang declaration, which is necessary for mapping a Clang struct parameter type to the Carbon class without doing name lookup.
To map the type, we also need to map namespaces. To avoid recursion for inner namespaces, we use a vector.

geoffromer commented on toolchain/check/import_cpp.cpp:
```diff
@@ -54,6 +54,11 @@ static auto GenerateCppIncludesHeaderCode(
 namespace {
 
+// Maps a Clang name to a Carbon `NameId`.
+static auto MapNameId(Context& context, llvm::StringRef name) -> SemIR::NameId {
```
What do you think of calling this `AddIdentifierName`?

bricknerb replied:
Done.

geoffromer commented on toolchain/sem_ir/clang_decl.h:
```diff
@@ -0,0 +1,59 @@
+};
+
+// Hashing for ClangDecl. See common/hashing.h.
+inline auto CarbonHashValue(const ClangDecl& value, uint64_t seed) -> HashCode {
```
Can this be a "hidden friend", like `CarbonHashtableEq`?

bricknerb replied:
Done.
FWIW, I was following the example of `EntityName`.
https://github.com/carbon-language/carbon-lang/blob/924705e6d8d93bcd1a074422bdf66e2f3527dce9/toolchain/sem_ir/entity_name.h#L22-L50

geoffromer replied:
I think that's just a historical accident -- equality and hashing were added in separate PRs by separate authors.

bricknerb replied:
SG. Sent out https://github.com/carbon-language/carbon-lang/pull/5701.

geoffromer commented on toolchain/sem_ir/clang_decl.h:
```diff
@@ -0,0 +1,59 @@
+}
+
+// The ID of a Clang declaration mapping, pointing to the Clang AST and the
+// mapped Carbon instruction.
```
```suggestion
// The ID of a `ClangDecl`.
```

Most of the rest of this is redundant with the documentation on `ClangDecl`.

bricknerb replied:
Done.

geoffromer commented on toolchain/sem_ir/clang_decl.cpp:
```diff
@@ -0,0 +1,17 @@
+auto ClangDecl::Print(llvm::raw_ostream& out) const -> void {
+  out << "{decl: ";
+  decl->print(out);
+  out << ", inst: " << inst_id << "}";
```
```suggestion
  out << ", inst_id: " << inst_id << "}";
```

bricknerb replied:
Done.

geoffromer commented on toolchain/check/import_cpp.cpp:
```diff
@@ -294,8 +300,11 @@ static auto ClangLookup(Context& context, SemIR::NameScopeId scope_id,
+      clang::dyn_cast<clang::DeclContext>(
+          context.sem_ir()
+              .clang_decls()
+              .Get(context.name_scopes().Get(scope_id).clang_decl_context_id())
```
Optional: this might read a little better if `context.name_scopes().Get(scope_id).clang_decl_context_id()` were pulled out as a named variable.

bricknerb replied:
Done.

geoffromer commented on toolchain/check/import_cpp.cpp:
```diff
@@ -304,6 +313,175 @@ static auto ClangLookup(Context& context, SemIR::NameScopeId scope_id,
+}
+
+// Maps a C++ declaration context to a Carbon namespace.
+static auto MapDeclContext(Context& context, clang::DeclContext* decl_context)
```
I find this name pretty unclear, because "Map" is vague, and "DeclContext" focuses on the type of the argument. When a function's return value is more important than its side effects (which I think is the case here?), it's usually clearer to name it after that return value, because that's the information that's most relevant at the callsite. What do you think of calling this something like `AsNamespace` or `AsCarbonNamespace`?

bricknerb replied:
Done.

geoffromer commented on toolchain/check/import_cpp.cpp:
```diff
@@ -304,6 +313,175 @@ static auto ClangLookup(Context& context, SemIR::NameScopeId scope_id,
+  auto& clang_decls = context.sem_ir().clang_decls();
+
+  // Check if the decl context is already mapped to a Carbon namespace.
+  auto context_clang_decl_id =
+      clang_decls.Lookup({.decl = clang::dyn_cast<clang::Decl>(decl_context),
+                          .inst_id = SemIR::InstId::None});
+  if (context_clang_decl_id.has_value()) {
+    return clang_decls.Get(context_clang_decl_id).inst_id;
+  }
```
Optional:

```suggestion
  if (auto context_clang_decl_id =
      clang_decls.Lookup({.decl = clang::dyn_cast<clang::Decl>(decl_context),
                          .inst_id = SemIR::InstId::None});
      context_clang_decl_id.has_value()) {
    return clang_decls.Get(context_clang_decl_id).inst_id;
  }
```

bricknerb replied:
Done. Thanks!

geoffromer commented on toolchain/sem_ir/file.h:
```diff
@@ -335,7 +335,8 @@ class File : public Printable<File> {
 
-  // Clang AST declarations pointing to the AST.
+  // Clang AST declarations pointing to the AST and their mapped Carbon
+  // instructions. When calling `Lookup()`, `inst_id` is ignored.
```
`inst_id` is also ignored when calling `Add` (but see my comment on `CarbonHashtableEq`).

bricknerb replied:
Done.

geoffromer commented on toolchain/sem_ir/clang_decl.h:
```diff
@@ -0,0 +1,59 @@
+struct ClangDecl : public Printable<ClangDecl> {
+  auto Print(llvm::raw_ostream& out) const -> void;
+
+  friend auto CarbonHashtableEq(const ClangDecl& lhs, const ClangDecl& rhs)
```
Hashing and equality comparison should operate on the full logical state of the object, and `inst_id` really seems like it's part of the logical state of `ClangDecl`. See also style guide, which [says](https://google.github.io/styleguide/cppguide.html#Operator_Overloading):

> ... don't define operator overloads just because other libraries expect them. For example, if your type doesn't have a natural ordering, but you want to store it in a `std::set`, use a custom comparator rather than overloading `<`."

I think the more general principle it's getting at is that fundamental operations like equality and hashing should be defined based on the meaning of the type as a whole, and not contorted to fit one particular use case. Excluding `inst_id` from hashing and equality feels like such a contortion to me.

Would it work to have a separate `Map<clang::Decl*, ClangDeclId>` for performing the lookup that you're trying to support here?

bricknerb replied:
The general principal makes sense to me, even though I'm not sure the style guide for operator overloads applies here.
I'm fine with a separate mapping, using a custom hashing was proposed [here](https://discord.com/channels/655572317891461132/768530752592805919/1380575881050718469), which also makes sense to me if it's considered an ok practice in Carbon toolchain. Sorry for not providing this context before. Feel free to continue the discussion in Discord or here. I'm fine with both approaches.

geoffromer replied:
> The general principal makes sense to me, even though I'm not sure the style guide for operator overloads applies here.

I agree that the style rule doesn't directly apply here; I was just citing it for the general principle.

> I'm fine with a separate mapping, using a custom hashing was proposed [here](https://discord.com/channels/655572317891461132/768530752592805919/1380575881050718469), which also makes sense to me if it's considered an ok practice in Carbon toolchain. Sorry for not providing this context before. Feel free to continue the discussion in Discord or here. I'm fine with both approaches.

As far as I know, this way of using a custom hash isn't an established practice in our codebase (at least, I can't find any existing examples of it). I've moved the discussion to Discord.

bricknerb replied:
Thanks, this all makes sense to me.
I've added a TODO to fix this in a follow up once we decide the best path forward.
Let me know if you prefer to block on this and which path do you think is best to proceed.

geoffromer commented on toolchain/check/import_cpp.cpp:
```diff
@@ -304,6 +313,175 @@ static auto ClangLookup(Context& context, SemIR::NameScopeId scope_id,
+  // We know the parent of the last decl context is mapped, map the rest.
+  auto namespace_inst_id = SemIR::InstId::None;
+  do {
+    decl_context = decl_contexts.back();
+    decl_contexts.pop_back();
```
```suggestion
    decl_context = decl_contexts.pop_back_val();
```

bricknerb replied:
Done.

geoffromer commented on toolchain/check/import_cpp.cpp:
```diff
@@ -304,6 +313,175 @@ static auto ClangLookup(Context& context, SemIR::NameScopeId scope_id,
+    decl_context = decl_contexts.back();
+    decl_contexts.pop_back();
+    auto parent_inst_id = clang_decls.Get(parent_decl_id).inst_id;
+    auto parent_namespace =
+        context.insts().Get(parent_inst_id).TryAs<SemIR::Namespace>();
+    CARBON_CHECK(parent_namespace.has_value());
```
```suggestion
    auto parent_namespace = 
        context.insts().GetAs<SemIR::Namespace>(parent_inst_id);
```

bricknerb replied:
Done.

geoffromer commented on toolchain/check/import_cpp.cpp:
```diff
@@ -339,6 +513,52 @@ static auto MapType(Context& context, clang::QualType type) -> TypeExpr {
+static auto MapRecordType(Context& context, SemIR::LocId loc_id,
+                          const clang::RecordType& type) -> TypeExpr {
+  auto* record_decl = clang::dyn_cast<clang::CXXRecordDecl>(type.getDecl());
+  if (record_decl && record_decl->isStruct()) {
```
Why does it matter whether the record is a struct? Given the marginal difference between structs and classes in C++, shouldn't they both get the same treatment?

bricknerb replied:
They will and we have a TODO.
Just trying to limit the scope of this change since we have structs, classes and unions and I want to make sure we consider the differences if any and make sure we cover everything in tests.

geoffromer commented on toolchain/check/import_cpp.cpp:
```diff
@@ -339,6 +513,52 @@ static auto MapType(Context& context, clang::QualType type) -> TypeExpr {
+          MapDeclContext(context, record_decl->getDeclContext());
+      auto parent_name_scope_id =
+          context.insts().GetAs<SemIR::Namespace>(parent_inst_id).name_scope_id;
+      SemIR::InstId struct_inst_id = ImportCXXRecordDecl(
```
Is there any risk of this call and the one in `ImportNameDecl` duplicating work, and importing the same class twice?

bricknerb replied:
Nice catch!

If `ImportNameDecl()` imports a struct, it will add it `clang_decls` so it won't be imported by `MapRecordType()`.
If `MapRecordType()` imports a struct, it indeed doesn't add its name, so we still import it when lookup by name afterwards.

To fix this:
1. I've removed the unnecessary attempt to add to `clang_decls` after calling `ImportCXXRecordDecl()`, which did nothing.
2. I've added tests.
3. I've added the name to the scope in this case.

bricknerb committed:
Use `GetAs()` instead of `Get()`...`TryAs()`...`CARBON_CHECK(...has_value())`

bricknerb committed:
Use `pop_back_val()` instead of `back()` and `pop_back()`

bricknerb committed:
Clarify `clang_decls_` cannot add multiple entries with the same `decl`

bricknerb committed:
Simplify `ClangDeclId` documentation

bricknerb committed:
Move `ClangDecl::CarbonHashValue()` to be a hidden friend

bricknerb committed:
Move initialization to the `if()`

bricknerb committed:
Rename `MapDeclContext()` to `AsCarbonNamespace()`

bricknerb committed:
Put `context.name_scopes().Get(scope_id).clang_decl_context_id()` in a variable

bricknerb committed:
Rename `MapNameId()` to `AddIdentifierName()`

bricknerb committed:
Change `ClangDecl::Print()` to write "inst_id" instead of "inst"

bricknerb committed:
Do not try to add the struct mapping to `clang_decls` since `ImportCXXRecordDecl()` does that on success

bricknerb committed:
Add tests for using a struct both in parameter and explicitly

bricknerb committed:
When mapping a struct, add its name to the name scope

bricknerb committed:
Move logic that adds imported name into scope to a new `ImportNameDeclIntoScope()` and use it by both `ImportNameFromCpp()` and `MapRecordType()`. This way that logic is deduplicated. However, this adds recursive call: `ImportNameFromCpp()` -> `ImportNameDeclIntoScope()` -> `ImportNameDecl()` -> `ImportFunctionDecl()` -> ... -> `MapType()` -> `MapRecordType()` -> `ImportNameDeclIntoScope()` -> `ImportNameDecl()` -> `ImportCXXRecordDecl()`.
Also take out static function from anonymous namespace.

bricknerb committed:
Add a TODO to avoid custom hashing in `ClangDecl`.

bricknerb committed:
Document `ImportNameDeclIntoScope()` and take out adding the name to a separate function, so we can reuse it instead of doing a recursive call.
